### PR TITLE
Bug fix tup may change if relation need oids.

### PR DESCRIPTION
**Pull request by ppggff**
## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by ashwinstar** [comment]
@ppggff Please can you add a test to demonstrate the problem and help validate the fix.

**Comment by ppggff** [comment]
> @ppggff Please can you add a test to demonstrate the problem and help validate the fix.

To reproduce this problem:
1. compile with `--enable-cassert`
2. run following sql:
```
set debug_appendonly_use_no_toast to on;

-- sql in qp_dml_oids.sql
CREATE TABLE dml_ao (a int , b int default -1, c text) WITH (appendonly = true, oids = true) DISTRIBUTED BY (a);
INSERT INTO dml_ao (a, b, c) VALUES (10, 1, repeat('x', 50000));
```

**Comment by ashwinstar** [comment]
yes please add those SQL to `qp_dml_oids.sql` as test. The code change looks good to me. Once test change is made, I will go ahead to commit the PR.

Just side notes `oids=true` might not be around for much longer time and also guc `debug_appendonly_use_no_toast` is sketchy I feel. I think we should just get rid of the same, as I don't think we expose the same as customer facing thing.

**Comment by ppggff** [comment]
@ashwinstar test added. Could you please check another pr https://github.com/greenplum-db/gpdb/pull/9486 ?

File: src/backend/access/appendonly/appendonlyam.c [hunk 1]
```diff
@@ -826,10 +826,13 @@ AppendOnlyExecutorReadBlock_Init(AppendOnlyExecutorReadBlock *executorReadBlock,
 {
 	MemoryContext oldcontext;
 
+	AssertArg(MemoryContextIsValid(memoryContext));
+
 	oldcontext = MemoryContextSwitchTo(memoryContext);
 	executorReadBlock->uncompressedBuffer = (uint8 *) palloc(usableBlockSize * sizeof(uint8));
 
 	executorReadBlock->storageRead = storageRead;
+	executorReadBlock->memoryContext = memoryContext;
```
Comment (ashwinstar) [inline src/backend/access/appendonly/appendonlyam.c]: Seems you added this new commit to PR. Was this exposed as well with added test?

Comment (ppggff) [reply src/backend/access/appendonly/appendonlyam.c]: Yes

**Comment by ashwinstar** [comment]
CI is failing currently since `expected/qp_dml_oids_optimizer.out` answer file also needs to be updated with the change. I will take care of that and push next week, need to build with orca and run the test. If you beat me on updating that answer file and pipeline flags green, that would be awesome as well.

**Comment by ppggff** [comment]
@ashwinstar update with `qp_dml_oids_optimizer.out`

**Comment by ashwinstar** [comment]
Pushed to master. Thanks!

**Pull request closed without merging**